File: misc/validate.py
#!/usr/bin/env python3
"""Validation suite for scraped data. Run after each scraper step to catch issues."""

import json
import re
import sys
import os
import html as html_lib
from dotenv import load_dotenv

load_dotenv()

DATABASE_URL = os.environ["DATABASE_URL"]

OUTPUT_DIR = "output"
FACEBOOK_FILE = os.path.join(OUTPUT_DIR, "students.json")
ENRICHED_FILE = os.path.join(OUTPUT_DIR, "students_enriched.json")

# --- Thresholds (adjust if Yale enrollment changes significantly) ---
MIN_TOTAL_STUDENTS = 6000
MAX_TOTAL_STUDENTS = 8000
MIN_ENRICHMENT_RATE = 0.90
MIN_COLLEGES = 14  # Yale has 14 residential colleges
EXPECTED_YEARS = {2026, 2027, 2028, 2029}  # Update each academic year
MIN_STUDENTS_PER_YEAR = 500
MIN_STUDENTS_PER_COLLEGE = 200

VALID_COLLEGES = {
    "Benjamin Franklin College", "Berkeley College", "Branford College",
    "Davenport College", "Ezra Stiles College", "Grace Hopper College",
    "Jonathan Edwards College", "Morse College", "Pauli Murray College",
    "Pierson College", "Saybrook College", "Silliman College",
    "Timothy Dwight College", "Trumbull College",
}

MONTH_NAMES = {"Jan", "Feb", "Mar", "Apr", "May", "Jun",
               "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}


class ValidationResult:
    def __init__(self):
        self.passes = []
        self.warnings = []
        self.failures = []

    def ok(self, msg):
        self.passes.append(msg)

    def warn(self, msg):
        self.warnings.append(msg)

    def fail(self, msg):
        self.failures.append(msg)

    def report(self, title):
        print(f"\n{'='*60}")
        print(f"  {title}")
        print(f"{'='*60}")
        for msg in self.passes:
            print(f"  PASS  {msg}")
        for msg in self.warnings:
            print(f"  WARN  {msg}")
        for msg in self.failures:
            print(f"  FAIL  {msg}")
        print(f"\n  Total: {len(self.passes)} passed, {len(self.warnings)} warnings, {len(self.failures)} failures")
        return len(self.failures) == 0


def load_json(path):
    if not os.path.exists(path):
        return None
    with open(path) as f:
        return json.load(f)


# --- Facebook Scraper Validation ---

def validate_facebook(data, result):
    """Validate output of scraper.py (students.json)."""

    # 1. Total count
    n = len(data)
    if MIN_TOTAL_STUDENTS <= n <= MAX_TOTAL_STUDENTS:
        result.ok(f"Student count: {n} (expected {MIN_TOTAL_STUDENTS}-{MAX_TOTAL_STUDENTS})")
    else:
        result.fail(f"Student count: {n} (expected {MIN_TOTAL_STUDENTS}-{MAX_TOTAL_STUDENTS})")

    # 2. Required fields present
    for field, label in [("full_name", "Name"), ("year", "Year"),
                         ("college", "College"), ("photo_id", "Photo ID")]:
        missing = sum(1 for s in data if not s.get(field))
        pct = missing / n * 100
        if pct == 0:
            result.ok(f"{label}: 100% populated")
        elif pct < 2:
            result.warn(f"{label}: {missing}/{n} missing ({pct:.1f}%)")
        else:
            result.fail(f"{label}: {missing}/{n} missing ({pct:.1f}%)")

    # 3. Optional fields coverage
    for field, label, min_pct in [("major", "Major", 80), ("birthday", "Birthday", 80),
                                   ("address", "Address", 70), ("phone", "Phone", 20)]:
        present = sum(1 for s in data if s.get(field))
        pct = present / n * 100
        if pct >= min_pct:
            result.ok(f"{label}: {present}/{n} ({pct:.1f}%) — above {min_pct}% threshold")
        else:
            result.warn(f"{label}: {present}/{n} ({pct:.1f}%) — below {min_pct}% threshold")

    # 4. College distribution
    colleges = {}
    for s in data:
        c = s.get("college", "MISSING")
        colleges[c] = colleges.get(c, 0) + 1

    known = {c for c in colleges if c in VALID_COLLEGES}
    unknown = {c for c in colleges if c not in VALID_COLLEGES and c != "MISSING"}

    if len(known) >= MIN_COLLEGES:
        result.ok(f"Colleges: {len(known)}/{MIN_COLLEGES} known residential colleges found")
    else:
        missing_colleges = VALID_COLLEGES - known
        result.fail(f"Colleges: only {len(known)}/{MIN_COLLEGES} — missing: {missing_colleges}")

    if unknown:
        result.warn(f"Unknown colleges found: {unknown}")

    for college, count in sorted(colleges.items()):
        if college in VALID_COLLEGES and count < MIN_STUDENTS_PER_COLLEGE:
            result.warn(f"  {college}: only {count} students (expected >={MIN_STUDENTS_PER_COLLEGE})")

    # 5. Year distribution
    years = {}
    for s in data:
        y = s.get("year", "")
        if y:
            match = re.match(r"'(\d{2})", y)
            if match:
                years[2000 + int(match.group(1))] = years.get(2000 + int(match.group(1)), 0) + 1

    for y in EXPECTED_YEARS:
        count = years.get(y, 0)
        if count >= MIN_STUDENTS_PER_YEAR:
            result.ok(f"Year {y}: {count} students")
        elif count > 0:
            result.warn(f"Year {y}: only {count} students (expected >={MIN_STUDENTS_PER_YEAR})")
        else:
            result.fail(f"Year {y}: 0 students found")

    unexpected_years = set(years.keys()) - EXPECTED_YEARS
    if unexpected_years:
        for y in sorted(unexpected_years):
            result.warn(f"Unexpected year {y}: {years[y]} students")

    # 6. Duplicate detection
    names = [s.get("full_name", "") for s in data]
    name_counts = {}
    for name in names:
        name_counts[name] = name_counts.get(name, 0) + 1
    dupes = {name: count for name, count in name_counts.items() if count > 1}
    if len(dupes) <= 20:
        result.ok(f"Duplicates: {len(dupes)} duplicate names (likely legitimate)")
    else:
        result.warn(f"Duplicates: {len(dupes)} duplicate names — top: {list(dupes.items())[:5]}")

    # 7. Data quality checks
    html_entities = sum(1 for s in data if "&amp;" in s.get("major", "") or "&lt;" in s.get("major", ""))
    if html_entities == 0:
        result.ok("HTML entities: none found in major field")
    else:
        result.fail(f"HTML entities: {html_entities} students have unescaped HTML in major")

    empty_first = sum(1 for s in data if not s.get("first_name"))
    if empty_first == 0:
        result.ok("First names: all populated")
    else:
        result.warn(f"First names: {empty_first} students have empty first_name")

    # 8. Photo IDs
    no_photo = sum(1 for s in data if s.get("photo_id") == "0")
    result.ok(f"Photos: {n - no_photo}/{n} have photos ({no_photo} with id=0)")


# --- Directory Enrichment Validation ---

def validate_enriched(data, result):
    """Validate output of directory_scraper.py (students_enriched.json)."""
    n = len(data)

    # 1. Enrichment rate
    with_netid = sum(1 for s in data if s.get("netid"))
    rate = with_netid / n
    if rate >= MIN_ENRICHMENT_RATE:
        result.ok(f"Enrichment rate: {with_netid}/{n} ({rate*100:.1f}%) — above {MIN_ENRICHMENT_RATE*100}% threshold")
    else:
        result.fail(f"Enrichment rate: {with_netid}/{n} ({rate*100:.1f}%) — below {MIN_ENRICHMENT_RATE*100}% threshold")

    # 2. Key enriched fields
    for field, label in [("netid", "NetID"), ("email", "Email"), ("upi", "UPI"),
                         ("school_code", "School Code"), ("college_code", "College Code"),
                         ("year_directory", "Grad Year"), ("mailbox", "Mailbox")]:
        present = sum(1 for s in data if s.get(field))
        pct = present / n * 100
        if pct >= MIN_ENRICHMENT_RATE * 100:
            result.ok(f"{label}: {present}/{n} ({pct:.1f}%)")
        elif pct >= 80:
            result.warn(f"{label}: {present}/{n} ({pct:.1f}%)")
        else:
            result.fail(f"{label}: {present}/{n} ({pct:.1f}%)")

    # 3. Email format validation
    bad_emails = []
    for s in data:
        email = s.get("email", "")
        if email and not re.match(r"^[^@]+@(.*\.)?yale\.edu$", email):
            bad_emails.append(email)
    if not bad_emails:
        result.ok("Email format: all emails match *@*.yale.edu")
    else:
        result.fail(f"Email format: {len(bad_emails)} invalid emails — e.g. {bad_emails[:3]}")

    # 4. NetID format validation
    bad_netids = []
    for s in data:
        netid = s.get("netid", "")
        if netid and not re.match(r"^[a-z]{2,4}\d{1,4}$", netid):
            bad_netids.append(netid)
    if not bad_netids:
        result.ok("NetID format: all netids match expected pattern")
    else:
        result.warn(f"NetID format: {len(bad_netids)} unusual netids — e.g. {bad_netids[:5]}")

    # 5. UPI format validation
    bad_upis = sum(1 for s in data if s.get("upi") and not isinstance(s["upi"], int))
    if bad_upis == 0:
        result.ok("UPI format: all UPIs are integers")
    else:
        result.fail(f"UPI format: {bad_upis} non-integer UPIs")

    # 6. Cross-field consistency
    mismatched_college = 0
    for s in data:
        fb_college = s.get("college", "").lower()
        dir_college = s.get("college_directory", "").lower()
        if fb_college and dir_college and fb_college != dir_college:
            mismatched_college += 1
    if mismatched_college == 0:
        result.ok("College consistency: facebook and directory colleges match")
    elif mismatched_college < 20:
        result.warn(f"College consistency: {mismatched_college} mismatches between facebook and directory")
    else:
        result.fail(f"College consistency: {mismatched_college} mismatches")

    # 7. Students not found
    not_enriched = [s for s in data if not s.get("netid")]
    if len(not_enriched) < n * 0.10:
        result.ok(f"Not found: {len(not_enriched)} students ({len(not_enriched)/n*100:.1f}%)")
    else:
        result.fail(f"Not found: {len(not_enriched)} students ({len(not_enriched)/n*100:.1f}%)")

    # Show some examples of not-found students
    if not_enriched:
        examples = [s.get("full_name", "?") for s in not_enriched[:5]]
        result.warn(f"Not-found examples: {examples}")


# --- Database Validation ---

def validate_database(database_url):
    """Validate the database state after loading."""
    try:
        import sqlalchemy
        from sqlalchemy import text as sql_text
    except ImportError:
        print("  SKIP  Database validation (sqlalchemy not installed)")
        return True

    result = ValidationResult()
    engine = sqlalchemy.create_engine(database_url)

    with engine.connect() as conn:
        # 1. Total counts
        total = conn.execute(sql_text("SELECT COUNT(*) FROM person")).scalar()
        yc = conn.execute(sql_text(
            "SELECT COUNT(*) FROM person WHERE school = 'Yale College'"
        )).scalar()
        result.ok(f"Total rows: {total}, Yale College: {yc}")

        # 2. Yale College count in range
        if MIN_TOTAL_STUDENTS <= yc <= MAX_TOTAL_STUDENTS:
            result.ok(f"YC count {yc} in expected range")
        else:
            result.fail(f"YC count {yc} outside expected range {MIN_TOTAL_STUDENTS}-{MAX_TOTAL_STUDENTS}")

        # 3. Null checks on required fields
        for col in ["first_name", "last_name"]:
            nulls = conn.execute(sql_text(
                f"SELECT COUNT(*) FROM person WHERE school = 'Yale College' AND {col} IS NULL"
            )).scalar()
            if nulls == 0:
                result.ok(f"DB {col}: no nulls")
            else:
                result.fail(f"DB {col}: {nulls} null values")

        # 4. NetID coverage
        with_netid = conn.execute(sql_text(
            "SELECT COUNT(*) FROM person WHERE school = 'Yale College' AND netid IS NOT NULL"
        )).scalar()
        rate = with_netid / yc if yc > 0 else 0
        if rate >= MIN_ENRICHMENT_RATE:
            result.ok(f"DB netid coverage: {with_netid}/{yc} ({rate*100:.1f}%)")
        else:
            result.fail(f"DB netid coverage: {with_netid}/{yc} ({rate*100:.1f}%)")

        # 5. Duplicate netids
        dup_netids = conn.execute(sql_text(
            "SELECT netid, COUNT(*) FROM person WHERE netid IS NOT NULL "
            "GROUP BY netid HAVING COUNT(*) > 1"
        )).fetchall()
        if not dup_netids:
            result.ok("DB netid uniqueness: no duplicates")
        else:
            result.fail(f"DB duplicate netids: {len(dup_netids)} — e.g. {dup_netids[:3]}")

        # 6. Duplicate IDs
        dup_ids = conn.execute(sql_text(
            "SELECT id, COUNT(*) FROM person GROUP BY id HAVING COUNT(*) > 1"
        )).fetchall()
        if not dup_ids:
            result.ok("DB id uniqueness: no duplicates")
        else:
            result.fail(f"DB duplicate ids: {len(dup_ids)}")

        # 7. College distribution
        colleges = conn.execute(sql_text(
            "SELECT college, COUNT(*) FROM person WHERE school = 'Yale College' "
            "GROUP BY college ORDER BY count DESC"
        )).fetchall()
        if len(colleges) >= MIN_COLLEGES:
            result.ok(f"DB colleges: {len(colleges)} distinct colleges")
        else:
            result.fail(f"DB colleges: only {len(colleges)}")

        # 8. Year distribution
        years = conn.execute(sql_text(
            "SELECT year, COUNT(*) FROM person WHERE school = 'Yale College' "
            "AND year IS NOT NULL GROUP BY year ORDER BY year"
        )).fetchall()
        for y, count in years:
            if y in EXPECTED_YEARS:
                result.ok(f"DB year {y}: {count} students")

        # 9. No orphaned data
        no_school = conn.execute(sql_text(
            "SELECT COUNT(*) FROM person WHERE school IS NULL AND school_code IS NULL "
            "AND organization IS NULL"
        )).scalar()
        if no_school == 0:
            result.ok("DB orphans: no records without school or organization")
        else:
            result.warn(f"DB orphans: {no_school} records with no school/org")

    return result.report("Database Validation")


def main():
    import argparse
    parser = argparse.ArgumentParser(description="Validate scraped data")
    parser.add_argument("--step", choices=["facebook", "enriched", "database", "all"],
                        default="all", help="Which step to validate")
    parser.add_argument("--database-url", type=str,
                        default=DATABASE_URL,
                        help="PostgreSQL connection URL for database validation")
    args = parser.parse_args()

    all_passed = True

    if args.step in ("facebook", "all"):
        data = load_json(FACEBOOK_FILE)
        if data:
            result = ValidationResult()
            validate_facebook(data, result)
            if not result.report("Facebook Scraper Validation (students.json)"):
                all_passed = False
        else:
            print(f"\n  SKIP  {FACEBOOK_FILE} not found")

    if args.step in ("enriched", "all"):
        data = load_json(ENRICHED_FILE)
        if data:
            result = ValidationResult()
            validate_facebook(data, result)  # Run base checks too
            if not result.report("Enriched Data — Base Checks (students_enriched.json)"):
                all_passed = False

            result2 = ValidationResult()
            validate_enriched(data, result2)
            if not result2.report("Enriched Data — Directory Checks (students_enriched.json)"):
                all_passed = False
        else:
            print(f"\n  SKIP  {ENRICHED_FILE} not found")

    if args.step in ("database", "all"):
        if not validate_database(args.database_url):
            all_passed = False

    print(f"\n{'='*60}")
    if all_passed:
        print("  ALL VALIDATIONS PASSED")
    else:
        print("  SOME VALIDATIONS FAILED — review above")
    print(f"{'='*60}")

    sys.exit(0 if all_passed else 1)


if __name__ == "__main__":
    main()

File: Yalies/yalies-internal/yalies-data-pipeline/src/validate.ts
import { Sequelize, QueryTypes } from "sequelize";
import { FacebookStudent, EnrichedStudent, ValidationResult } from "./types.js";
import {
	VALID_COLLEGES,
	EXPECTED_YEARS,
	YALE_EMAIL_REGEX,
	NETID_REGEX,
	VALIDATION_THRESHOLDS,
	YALE_COLLEGE,
} from "yalies-shared";

const {
	MIN_TOTAL_STUDENTS,
	MAX_TOTAL_STUDENTS,
	MIN_ENRICHMENT_RATE,
	MIN_COLLEGES,
	MIN_STUDENTS_PER_YEAR,
	MIN_STUDENTS_PER_COLLEGE,
} = VALIDATION_THRESHOLDS;

const VALID_COLLEGES_SET = new Set<string>(VALID_COLLEGES);
const EXPECTED_YEARS_SET = new Set<number>(EXPECTED_YEARS);

function createResult(): ValidationResult {
	return { passes: [], warnings: [], failures: [] };
}

function reportResult(result: ValidationResult, title: string): boolean {
	console.log(`\n${"=".repeat(60)}`);
	console.log(`  ${title}`);
	console.log(`${"=".repeat(60)}`);
	for (const msg of result.passes) console.log(`  PASS  ${msg}`);
	for (const msg of result.warnings) console.log(`  WARN  ${msg}`);
	for (const msg of result.failures) console.log(`  FAIL  ${msg}`);
	console.log(`\n  Total: ${result.passes.length} passed, ${result.warnings.length} warnings, ${result.failures.length} failures`);
	return result.failures.length === 0;
}

export function validateFacebook(data: FacebookStudent[]): ValidationResult {
	const result = createResult();
	const n = data.length;

	// 1. Total count
	if (n >= MIN_TOTAL_STUDENTS && n <= MAX_TOTAL_STUDENTS) {
		result.passes.push(`Student count: ${n} (expected ${MIN_TOTAL_STUDENTS}-${MAX_TOTAL_STUDENTS})`);
	} else {
		result.failures.push(`Student count: ${n} (expected ${MIN_TOTAL_STUDENTS}-${MAX_TOTAL_STUDENTS})`);
	}

	// 2. Required fields
	for (const [field, label] of [["full_name", "Name"], ["year", "Year"], ["college", "College"], ["photo_id", "Photo ID"]] as const) {
		const missing = data.filter((s) => !s[field]).length;
		const pct = (missing / n) * 100;
		if (pct === 0) result.passes.push(`${label}: 100% populated`);
		else if (pct < 2) result.warnings.push(`${label}: ${missing}/${n} missing (${pct.toFixed(1)}%)`);
		else result.failures.push(`${label}: ${missing}/${n} missing (${pct.toFixed(1)}%)`);
	}

	// 3. Optional fields
	for (const [field, label, minPct] of [["major", "Major", 80], ["birthday", "Birthday", 80], ["address", "Address", 70], ["phone", "Phone", 20]] as const) {
		const present = data.filter((s) => s[field as keyof FacebookStudent]).length;
		const pct = (present / n) * 100;
		if (pct >= (minPct as number)) {
			result.passes.push(`${label}: ${present}/${n} (${pct.toFixed(1)}%) — above ${minPct}% threshold`);
		} else {
			result.warnings.push(`${label}: ${present}/${n} (${pct.toFixed(1)}%) — below ${minPct}% threshold`);
		}
	}

	// 4. College distribution
	const colleges: Record<string, number> = {};
	for (const s of data) {
		const c = s.college || "MISSING";
		colleges[c] = (colleges[c] || 0) + 1;
	}

	const knownColleges = Object.keys(colleges).filter((c) => VALID_COLLEGES_SET.has(c));
	const unknownColleges = Object.keys(colleges).filter((c) => !VALID_COLLEGES_SET.has(c) && c !== "MISSING");

	if (knownColleges.length >= MIN_COLLEGES) {
		result.passes.push(`Colleges: ${knownColleges.length}/${MIN_COLLEGES} known residential colleges found`);
	} else {
		const missing = [...VALID_COLLEGES_SET].filter((c) => !knownColleges.includes(c));
		result.failures.push(`Colleges: only ${knownColleges.length}/${MIN_COLLEGES} — missing: ${missing.join(", ")}`);
	}

	if (unknownColleges.length > 0) {
		result.warnings.push(`Unknown colleges found: ${unknownColleges.join(", ")}`);
	}

	for (const [college, count] of Object.entries(colleges)) {
		if (VALID_COLLEGES_SET.has(college) && count < MIN_STUDENTS_PER_COLLEGE) {
			result.warnings.push(`  ${college}: only ${count} students (expected >=${MIN_STUDENTS_PER_COLLEGE})`);
		}
	}

	// 5. Year distribution
	const years: Record<number, number> = {};
	for (const s of data) {
		const match = s.year?.match(/^'(\d{2})$/);
		if (match) {
			const y = 2000 + parseInt(match[1]);
			years[y] = (years[y] || 0) + 1;
		}
	}

	for (const y of EXPECTED_YEARS_SET) {
		const count = years[y] || 0;
		if (count >= MIN_STUDENTS_PER_YEAR) result.passes.push(`Year ${y}: ${count} students`);
		else if (count > 0) result.warnings.push(`Year ${y}: only ${count} students (expected >=${MIN_STUDENTS_PER_YEAR})`);
		else result.failures.push(`Year ${y}: 0 students found`);
	}

	for (const [y, count] of Object.entries(years)) {
		if (!EXPECTED_YEARS_SET.has(parseInt(y))) {
			result.warnings.push(`Unexpected year ${y}: ${count} students`);
		}
	}

	// 6. Duplicates
	const nameCounts: Record<string, number> = {};
	for (const s of data) {
		const name = s.full_name || "";
		nameCounts[name] = (nameCounts[name] || 0) + 1;
	}
	const dupes = Object.entries(nameCounts).filter(([, c]) => c > 1);
	if (dupes.length <= 20) {
		result.passes.push(`Duplicates: ${dupes.length} duplicate names (likely legitimate)`);
	} else {
		const top = dupes.slice(0, 5).map(([name, count]) => `${name}(${count})`).join(", ");
		result.warnings.push(`Duplicates: ${dupes.length} duplicate names — top: ${top}`);
	}

	// 7. HTML entities
	const htmlEntities = data.filter((s) => s.major && (s.major.includes("&amp;") || s.major.includes("&lt;"))).length;
	if (htmlEntities === 0) result.passes.push("HTML entities: none found in major field");
	else result.failures.push(`HTML entities: ${htmlEntities} students have unescaped HTML in major`);

	// 8. Photos
	const noPhoto = data.filter((s) => s.photo_id === "0").length;
	result.passes.push(`Photos: ${n - noPhoto}/${n} have photos (${noPhoto} with id=0)`);

	return result;
}

export function validateEnriched(data: EnrichedStudent[]): ValidationResult {
	const result = createResult();
	const n = data.length;

	// 1. Enrichment rate
	const withNetid = data.filter((s) => s.netid).length;
	const rate = withNetid / n;
	if (rate >= MIN_ENRICHMENT_RATE) {
		result.passes.push(`Enrichment rate: ${withNetid}/${n} (${(rate * 100).toFixed(1)}%) — above ${MIN_ENRICHMENT_RATE * 100}% threshold`);
	} else {
		result.failures.push(`Enrichment rate: ${withNetid}/${n} (${(rate * 100).toFixed(1)}%) — below ${MIN_ENRICHMENT_RATE * 100}% threshold`);
	}

	// 2. Key enriched fields
	const enrichedFields: [string, string][] = [
		["netid", "NetID"], ["email", "Email"], ["upi", "UPI"],
		["school_code", "School Code"], ["college_code", "College Code"],
		["year_directory", "Grad Year"], ["mailbox", "Mailbox"],
	];

	for (const [field, label] of enrichedFields) {
		const present = data.filter((s) => (s as Record<string, unknown>)[field]).length;
		const pct = (present / n) * 100;
		if (pct >= MIN_ENRICHMENT_RATE * 100) result.passes.push(`${label}: ${present}/${n} (${pct.toFixed(1)}%)`);
		else if (pct >= 80) result.warnings.push(`${label}: ${present}/${n} (${pct.toFixed(1)}%)`);
		else result.failures.push(`${label}: ${present}/${n} (${pct.toFixed(1)}%)`);
	}

	// 3. Email format
	const badEmails = data.filter((s) => s.email && !YALE_EMAIL_REGEX.test(s.email));
	if (badEmails.length === 0) result.passes.push("Email format: all emails match *@*.yale.edu");
	else result.failures.push(`Email format: ${badEmails.length} invalid emails — e.g. ${badEmails.slice(0, 3).map((s) => s.email).join(", ")}`);

	// 4. NetID format
	const badNetids = data.filter((s) => s.netid && !NETID_REGEX.test(s.netid));
	if (badNetids.length === 0) result.passes.push("NetID format: all netids match expected pattern");
	else result.warnings.push(`NetID format: ${badNetids.length} unusual netids — e.g. ${badNetids.slice(0, 5).map((s) => s.netid).join(", ")}`);

	// 5. UPI format
	const badUpis = data.filter((s) => s.upi && typeof s.upi !== "number").length;
	if (badUpis === 0) result.passes.push("UPI format: all UPIs are integers");
	else result.failures.push(`UPI format: ${badUpis} non-integer UPIs`);

	// 6. College consistency
	const mismatched = data.filter((s) => {
		const fb = s.college?.toLowerCase() || "";
		const dir = s.college_directory?.toLowerCase() || "";
		return fb && dir && fb !== dir;
	}).length;
	if (mismatched === 0) result.passes.push("College consistency: facebook and directory colleges match");
	else if (mismatched < 20) result.warnings.push(`College consistency: ${mismatched} mismatches`);
	else result.failures.push(`College consistency: ${mismatched} mismatches`);

	// 7. Not-found students
	const notEnriched = data.filter((s) => !s.netid);
	if (notEnriched.length < n * 0.10) {
		result.passes.push(`Not found: ${notEnriched.length} students (${((notEnriched.length / n) * 100).toFixed(1)}%)`);
	} else {
		result.failures.push(`Not found: ${notEnriched.length} students (${((notEnriched.length / n) * 100).toFixed(1)}%)`);
	}

	if (notEnriched.length > 0) {
		const examples = notEnriched.slice(0, 5).map((s) => s.full_name).join(", ");
		result.warnings.push(`Not-found examples: ${examples}`);
	}

	return result;
}

export async function validateDatabase(databaseUrl: string): Promise<ValidationResult> {
	const result = createResult();
	const sequelize = new Sequelize(databaseUrl, { logging: false });

	try {
		await sequelize.authenticate();

		const query = async (sql: string) => {
			const [rows] = await sequelize.query<{ count: string }>(sql, { type: QueryTypes.SELECT });
			return parseInt(rows.count);
		};

		// 1. Total counts
		const total = await query("SELECT COUNT(*) as count FROM person");
		const yc = await query(`SELECT COUNT(*) as count FROM person WHERE school = '${YALE_COLLEGE}'`);
		result.passes.push(`Total rows: ${total}, ${YALE_COLLEGE}: ${yc}`);

		// 2. YC count in range
		if (yc >= MIN_TOTAL_STUDENTS && yc <= MAX_TOTAL_STUDENTS) {
			result.passes.push(`YC count ${yc} in expected range`);
		} else {
			result.failures.push(`YC count ${yc} outside expected range ${MIN_TOTAL_STUDENTS}-${MAX_TOTAL_STUDENTS}`);
		}

		// 3. Null checks
		for (const col of ["first_name", "last_name"]) {
			const nulls = await query(`SELECT COUNT(*) as count FROM person WHERE school = '${YALE_COLLEGE}' AND ${col} IS NULL`);
			if (nulls === 0) result.passes.push(`DB ${col}: no nulls`);
			else result.failures.push(`DB ${col}: ${nulls} null values`);
		}

		// 4. NetID coverage
		const withNetid = await query(`SELECT COUNT(*) as count FROM person WHERE school = '${YALE_COLLEGE}' AND netid IS NOT NULL`);
		const rate = withNetid / yc;
		if (rate >= MIN_ENRICHMENT_RATE) {
			result.passes.push(`DB netid coverage: ${withNetid}/${yc} (${(rate * 100).toFixed(1)}%)`);
		} else {
			result.failures.push(`DB netid coverage: ${withNetid}/${yc} (${(rate * 100).toFixed(1)}%)`);
		}

		// 5. Duplicate netids
		const [dupNetids] = await sequelize.query(
			"SELECT netid, COUNT(*) as count FROM person WHERE netid IS NOT NULL GROUP BY netid HAVING COUNT(*) > 1",
		);
		if ((dupNetids as unknown[]).length === 0) result.passes.push("DB netid uniqueness: no duplicates");
		else result.failures.push(`DB duplicate netids: ${(dupNetids as unknown[]).length}`);

		// 6. Duplicate IDs
		const [dupIds] = await sequelize.query(
			"SELECT id, COUNT(*) as count FROM person GROUP BY id HAVING COUNT(*) > 1",
		);
		if ((dupIds as unknown[]).length === 0) result.passes.push("DB id uniqueness: no duplicates");
		else result.failures.push(`DB duplicate ids: ${(dupIds as unknown[]).length}`);

		// 7. College distribution
		const [colleges] = await sequelize.query(
			`SELECT college, COUNT(*) as count FROM person WHERE school = '${YALE_COLLEGE}' GROUP BY college ORDER BY count DESC`,
		);
		if ((colleges as unknown[]).length >= MIN_COLLEGES) {
			result.passes.push(`DB colleges: ${(colleges as unknown[]).length} distinct colleges`);
		} else {
			result.failures.push(`DB colleges: only ${(colleges as unknown[]).length}`);
		}

		// 8. Year distribution
		const [yearRows] = await sequelize.query(
			`SELECT year, COUNT(*) as count FROM person WHERE school = '${YALE_COLLEGE}' AND year IS NOT NULL GROUP BY year ORDER BY year`,
		);
		for (const row of yearRows as { year: number; count: string }[]) {
			if (EXPECTED_YEARS_SET.has(row.year)) {
				result.passes.push(`DB year ${row.year}: ${row.count} students`);
			}
		}

		// 9. Orphaned data
		const orphans = await query(
			"SELECT COUNT(*) as count FROM person WHERE school IS NULL AND school_code IS NULL AND organization IS NULL",
		);
		if (orphans === 0) result.passes.push("DB orphans: no records without school or organization");
		else result.warnings.push(`DB orphans: ${orphans} records with no school/org`);

	} finally {
		await sequelize.close();
	}

	return result;
}

export async function runValidation(
	step: "facebook" | "enriched" | "database" | "all",
	data?: {
		facebook?: FacebookStudent[];
		enriched?: EnrichedStudent[];
		databaseUrl?: string;
	},
): Promise<boolean> {
	let allPassed = true;

	if ((step === "facebook" || step === "all") && data?.facebook) {
		const result = validateFacebook(data.facebook);
		if (!reportResult(result, "Facebook Scraper Validation")) allPassed = false;
	}

	if ((step === "enriched" || step === "all") && data?.enriched) {
		const baseResult = validateFacebook(data.enriched);
		if (!reportResult(baseResult, "Enriched Data — Base Checks")) allPassed = false;

		const enrichResult = validateEnriched(data.enriched);
		if (!reportResult(enrichResult, "Enriched Data — Directory Checks")) allPassed = false;
	}

	if ((step === "database" || step === "all") && data?.databaseUrl) {
		const dbResult = await validateDatabase(data.databaseUrl);
		if (!reportResult(dbResult, "Database Validation")) allPassed = false;
	}

	console.log(`\n${"=".repeat(60)}`);
	if (allPassed) {
		console.log("  ALL VALIDATIONS PASSED");
	} else {
		console.log("  SOME VALIDATIONS FAILED — review above");
	}
	console.log("=".repeat(60));

	return allPassed;
}
